src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AQNBuilder.Mono.cs

```diff
@@ -0,0 +1,483 @@
+				{
+					m_pStr.Remove(m_pStr.Length - 1, 1);
+				}
+				else
+				{
+					if (m_bUseAngleBracketsForGenerics)
+						Append('>');
+					else
+						Append(']');
+				}
+			}
+			public void OpenGenericArgument()
+			{
+				if(!CheckParseState (ParseState.START))
+					Fail ();
+
+				if (m_instNesting == 0)
+					Fail ();
+
+				m_parseState = ParseState.START;
+				m_bNestedName = false;
+
+				if (!m_bFirstInstArg)
+					Append(',');
+
+				m_bFirstInstArg = false;
+
+				if (m_bUseAngleBracketsForGenerics)
+					Append('<');
+				else
+					Append('[');
+				PushOpenGenericArgument();
+			}
+
+			public void CloseGenericArgument()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR | ParseState.BYREF | ParseState.ASSEMSPEC))
+					Fail();
+
+				if (m_instNesting == 0)
+					Fail();
+
+				m_parseState = ParseState.START;
+
+				if (m_bHasAssemblySpec)
+				{
+					if (m_bUseAngleBracketsForGenerics)
+						Append('>');
+					else
+						Append(']');
+				}
+
+				PopOpenGenericArgument();
+			}
+			public void AddName(string szName)
+			{
+				if (szName == null)
+					Fail ();
+
+				if (!CheckParseState(ParseState.START | ParseState.NAME))
+					Fail ();
+
+				m_parseState = ParseState.NAME;
+
+				if (m_bNestedName)
+					Append ('+');
+
+				m_bNestedName = true;
+
+				EscapeName (szName);
+			}
+
+			public void AddPointer()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail ();
+
+				m_parseState = ParseState.PTRARR;
+
+				Append ('*');
+			}
+
+			public void AddByRef()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.BYREF;
+
+				Append('&');
+			}
+
+			public void AddSzArray()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.PTRARR;
+
+				Append("[]");
+			}
+			public void AddArray(int rank)
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.PTRARR;
+
+				if (rank <= 0)
+					throw new ArgumentOutOfRangeException ();
+
+				if (rank == 1)
+					Append("[*]");
+				else if (rank > 64)
+				{
+					// Only taken in an error path, runtime will not load arrays of more than 32 dimentions
+					Append ($"[{rank}]");
+				}
+				else
+				{
+					var strDim = new StringBuilder (rank + 3);
+
+					strDim.Append ('[');
+					for(int i = 1; i < rank; i++) strDim.Append (',');
+					strDim.Append (']');
+					Append(strDim.ToString ());
+				}
+			}
+			public void AddAssemblySpec(string assemblySpec)
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR | ParseState.BYREF))
+					Fail();
+
+				m_parseState = ParseState.ASSEMSPEC;
+
+				if (assemblySpec != null && !assemblySpec.Equals(String.Empty))
+				{
+					Append(", ");
```

Comment:
This append helper is appending character by character to the underlying StringBuilder, and the only additional work it does is looking for a '\0' and stopping when it finds one.  In a case like this, we're passing in a string we know doesn't contain nulls.  Seems like in general we can just skip the helper and append directly to the string builder, only using the helper when dealing with unknown inputs, in which case it can also be named AppendUntilNull, or something like that.

Reply:
@stephentoub I was surprised that type names with nulls could be given to TypeBuilder* (the original reason for porting this to Mono), and I am not sure that e.g. Mono's `AssemblyBuilder` guarantees the same. A quick look at Mono's `AssemblyName` and `AssemblyBuilder` doesn't  preclude it - only the first character seems to get checked for null equivalence.

*(Importantly, the string passed to TypeBuilder is kept and accessible with the `Name` property, nulls and all; it is the `FullName` representation that must not have nulls. There are tests for this behavior)